### PR TITLE
WIP: Add squad metric

### LINKED ISSUE #60: Integrate SQuAD v2 metric
## 🚀 Feature

Currently there is no way to calculate the SQuAD metric. This is due to some complications in the metric aggregation. A wip branch can be seen here: https://github.com/PyTorchLightning/lightning-transformers/tree/wip-qa-metric

Continue integration of a QA metric!

### PR DESCRIPTION
**Pull request by benschreiber**
Partially resolves #60 (this PR adds squad v1, but v2 should be easy to add)

### CODE REVIEW

File: lightning_transformers/task/nlp/question_answering/datasets/squad/metric.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+class SquadMetric(Metric):
+
+    def __init__(self, postprocess_func, example_id_strings):
+        super().__init__(compute_on_step=False)
+        self.metric = load_metric("squad")
+        self.postprocess_func = postprocess_func
+        self.example_id_strings = example_id_strings
+        self.add_state("start_logits", [])
+        self.add_state("end_logits", [])
+        self.add_state("example_ids", [])
```
Comment (benschreiber) [inline lightning_transformers/task/nlp/question_answering/datasets/squad/metric.py:15]: I'm not sure if we need to store `example_ids`. Currently, the code assumes that the outputs are in the same order as the inputs. If that assumption can break (e.g. if update ordering is not preserved during distributed training), then we may need to sort by `example_ids` before passing the predictions to `postprocess_func` in `compute`.

File: lightning_transformers/task/nlp/question_answering/model.py [hunk 2]
```diff
@@ -43,3 +46,29 @@ def training_step(self, batch: Any, batch_idx: int) -> torch.Tensor:
...
+    def validation_step(self, batch: Any, batch_idx: int, dataloader_idx: int = 0) -> Dict[str, torch.Tensor]:
+        batch.pop("offset_mapping")
+        example_ids = batch.pop("example_id")
+        outputs = self.model(**batch)
+        loss = outputs[0]
+        self.log("val_loss", loss, prog_bar=True, sync_dist=True)
+        self.metric(example_ids, outputs.start_logits, outputs.end_logits)
+        return loss
+
+    def validation_epoch_end(self, outputs: Any) -> None:
```
Comment (benschreiber) [inline lightning_transformers/task/nlp/question_answering/model.py:59]: The metric is only computed over the entire dataset. I wasn't sure where to put that, but this seemed right.

Comment (SeanNaren) [reply lightning_transformers/task/nlp/question_answering/model.py:59]: Torchmetrics automatically get called `update` on every step and `compute` at the end through the logger, so this isn't needed I think! I will test

File: lightning_transformers/task/nlp/question_answering/datasets/squad/metric.py (hunk 1 above)
Comment (SeanNaren) [reply lightning_transformers/task/nlp/question_answering/datasets/squad/metric.py:15]: Great point, I'll need to read through the implementation, but I think this was also the case when I first looked at the squad v2 metric!

**Comment by SeanNaren** [comment]
This is awesome @benschreiber thanks for contributing this :) I can have a go testing once you feel it's ready.

**Comment by benschreiber** [comment]
Thanks @SeanNaren. Yeah, it's ready to test.

File: lightning_transformers/task/nlp/question_answering/model.py (hunk 2 above)
Comment (benschreiber) [reply lightning_transformers/task/nlp/question_answering/model.py:59]: It looked like `compute` was only being called through `forward`. I initialize the `Metric` with `compute_on_step=False` (the code is not equipped to compute the metric over a partial dataset, but you might be able to modify it to permit that), which makes `forward` only call `update`, not `compute`. I'm not sure if `compute` gets explicitly called anywhere. So yes, please test.

Comment (SeanNaren) [reply lightning_transformers/task/nlp/question_answering/model.py:59]: Asked someone on the torchmetrics team, your implementation is correct since it's a dict output

**Comment by SeanNaren** [comment]
Hey @benschreiber made a small modification to the validation step. I'm noticing for the test it's not collecting enough example ids in the lookup table. I can assist in debugging this but if you get a chance to have a look would appreciate it!

**Comment by benschreiber** [comment]
Hi @SeanNaren, I'm not sure what you are referring to. I just fixed an issue where metric state was not being reset between validation epochs. This would cause metric computation to fail the second time. This would manifest as an assert where `len(predictions[0]) == 21644` instead of 10822. 

It's also necessary to launch with `trainer.num_sanity_val_steps=-1` or `0` (the entire val dataset or nothing). Omitting this flag fails at the same assert, but with `len(predictions[0]) == 32`. Maybe we could check how many samples we have seen at the end of the val epoch and skip the metric computation if it's the wrong number.

Are either of these the bug that you saw?

**Comment by benschreiber** [comment]
Hi @SeanNaren, what's your current status on this?

**Review by SeanNaren — APPROVED** [review]
LGTM Thanks @benschreiber!

Testing failing is unrelated, needs to be fixed in a separate PR
